From 5a2457f967403b15ecfd30d5ce57d85994930303 Mon Sep 17 00:00:00 2001
From: Etaash Mathamsetty
 <45927311+Etaash-mathamsetty@users.noreply.github.com>
Date: Tue, 11 Aug 2026 15:15:38 -0400
Subject: [PATCH 269/285] winewayland: Use common EDID generation in win32u.

At this point this driver might as well be the ship of Theseus.
---
 dlls/winewayland.drv/display.c        | 184 +++++---------------------
 dlls/winewayland.drv/wayland_output.c |  56 +++-----
 dlls/winewayland.drv/waylanddrv.h     |  12 +-
 3 files changed, 65 insertions(+), 187 deletions(-)

diff --git a/dlls/winewayland.drv/display.c b/dlls/winewayland.drv/display.c
index f52b4ed71f4..f47e1e092f0 100644
--- a/dlls/winewayland.drv/display.c
+++ b/dlls/winewayland.drv/display.c
@@ -272,157 +272,9 @@ static void wayland_add_device_source(const struct gdi_device_manager *device_ma
     device_manager->add_source(output->name, state_flags, dpi, param);
 }
 
-/* borrowed from gamescope with permission */
-static uint8_t encode_max_luminance(float nits)
+static inline void ascii_to_unicode(WCHAR *dst, const char *src, int len)
 {
-    if (nits == 0.0f)
-        return 0;
-
-    return ceilf((logf(nits / 50.0f) / logf(2.0f)) * 32.0f);
-}
-
-static UINT get_edid(struct wayland_output_state *output, unsigned char **edid)
-{
-    const struct wayland_primaries *primaries = &output->primaries;
-    struct wayland_output_mode *mode = output->current_mode;
-    const char *model = output->model;
-    unsigned int edid_size = 128, extensions = 0;
-    unsigned char l[3] = {19, 1, 13}; /* SAM */
-    unsigned int i, mwidth, mheight;
-    unsigned char *data, *p, c;
-    char temp_model[13] = {0};
-
-    if (output->supports_hdr)
-    {
-        edid_size += 128;
-        extensions++;
-    }
-
-    if (!(*edid = calloc(edid_size, sizeof(**edid))))
-        return 0;
-
-    data = *edid;
-
-    mwidth = output->physical_w;
-    mheight = output->physical_h;
-
-    if (mwidth == 0 || mheight == 0)
-    {
-        /* assume ~150 dpi */
-        mwidth = mode->width / 60;
-        mheight = mode->width / 60;
-    }
-
-    *(uint64_t*)data = 0x00ffffffffffff00;
-
-    /* we cannot get this information from wayland, so make something up */
-    data[8] = ((l[0] & 0x1f) << 2) | ((l[1] & 0x18) >> 3);
-    data[9] = ((l[1] & 0x7) << 5) | (l[2] & 0x1f);
-    data[10] = 0xad;
-    data[11] = 0xde;
-    /* serial number is all zeros */
-    data[16] = 0xFF;
-    data[17] = 31; /* 2021 */
-    data[18] = 1;
-    data[19] = 4;
-    data[20] = 0xa0; /* FIXME */
-    data[21] = round(mwidth / 10.0); /* cm */
-    data[22] = round(mheight / 10.0); /* cm */
-    data[24] = 0x6;
-    data[25] = ((primaries->r_x & 0x3) << 6) | ((primaries->r_y & 0x3) << 4) |
-               ((primaries->g_x & 0x3) << 2) | (primaries->g_y & 0x3);
-    data[26] = ((primaries->b_x & 0x3) << 6) | ((primaries->b_y & 0x3) << 4) |
-               ((primaries->w_x & 0x3) << 2) | (primaries->w_y & 0x3);
-    data[27] = (primaries->r_x & 0x3fc) >> 2;
-    data[28] = (primaries->r_y & 0x3fc) >> 2;
-    data[29] = (primaries->g_x & 0x3fc) >> 2;
-    data[30] = (primaries->g_y & 0x3fc) >> 2;
-    data[31] = (primaries->b_x & 0x3fc) >> 2;
-    data[32] = (primaries->b_y & 0x3fc) >> 2;
-    data[33] = (primaries->w_x & 0x3fc) >> 2;
-    data[34] = (primaries->w_y & 0x3fc) >> 2;
-
-    for (i = 0; i < 16; ++i) data[38 + i] = 1;
-
-    p = data + 54;
-
-    *(uint16_t*)p = 0x0; /* 0 = reserved */
-
-    /* assume blanking time is 0 */
-    p[2] = mode->width;
-    p[4] = (((mode->width >> 8) & 0xf) << 4);
-    p[5] = mode->height;
-    p[7] = (((mode->height >> 8) & 0xf) << 4);
-    p[12] = mwidth;
-    p[13] = mheight;
-    p[14] = (((mwidth >> 8) & 0xf) << 4) | ((mheight >> 8) & 0xf);
-
-    p += 18;
-    p[3] = 0xfc;
-
-    if (model) lstrcpynA(temp_model, model, sizeof(temp_model));
-    else strcpy(temp_model, "Default");
-
-    for (i = 0; i < sizeof(temp_model); i++)
-    {
-        if (!temp_model[i])
-        {
-            temp_model[i++] = '\n';
-            break;
-        }
-    }
-    for (; i < sizeof(temp_model); i++)
-    {
-        if (!temp_model[i]) temp_model[i] = ' ';
-    }
-
-    TRACE("edid model %s\n", debugstr_an(temp_model, sizeof(temp_model)));
-    memcpy((char *)p + 5, temp_model, sizeof(temp_model));
-
-    p += 18;
-    p[3] = 0x10;
-    p += 18;
-    p[3] = 0x10;
-
-    c = 0;
-    data[126] = extensions;
-    for (i = 0; i < 127; ++i)
-        c += data[i];
-    data[127] = 256 - c;
-
-    p = data;
-
-    if (output->supports_hdr)
-    {
-        p += 128;
-
-        p[0] = 2;
-        p[1] = 3;
-        p[2] = 0xb;
-
-        p += 4;
-
-        p[0] = (0x7 << 5) | 0x6; /* HDR static metadata size */
-        p[1] = 6;
-
-        /* HDR static metadata block */
-
-        p[2] = 0x7; /* ST2084 | SDR | HDR */
-        p[3] = 1;
-        p[4] = encode_max_luminance(output->max_cll);
-        p[5] = encode_max_luminance(output->max_fall);
-        p[6] = 0; /* assume undefined, games often don't implement this properly */
-
-        /* reset p to beginning of the CTA block */
-        p = data + 128;
-        c = 0;
-
-        for (i = 0; i < 127; ++i)
-            c += p[i];
-        p[127] = 256 - c;
-    }
-
-    return edid_size;
+    while (len-- && *src) *dst++ = (unsigned char)*src++;
 }
 
 static void wayland_add_device_monitor(const struct gdi_device_manager *device_manager,
@@ -431,17 +283,47 @@ static void wayland_add_device_monitor(const struct gdi_device_manager *device_m
 {
     const char *env;
     struct gdi_monitor monitor = {0};
+    const unsigned int cta_mask = WAYLAND_OUTPUT_FALL | WAYLAND_OUTPUT_CLL;
+    struct edid_monitor_info *info = &monitor.edid_info;
 
     SetRect(&monitor.rc_monitor, output->physical_x, output->physical_y,
             output->physical_x + output->current_mode->width,
             output->physical_y + output->current_mode->height);
     OffsetRect(&monitor.rc_monitor, -primary->physical_x, -primary->physical_y);
 
-    monitor.edid_len = get_edid(output, &monitor.edid);
     /* We don't have a direct way to get the work area in Wayland. */
     monitor.rc_work = monitor.rc_monitor;
     monitor.hdr_enabled = output->supports_hdr;
 
+    info->flags = MONITOR_INFO_HAS_PRIMARIES | MONITOR_INFO_HAS_PREFERRED_MODE
+                  | MONITOR_INFO_HAS_PHYSICAL_DIMENSIONS | MONITOR_INFO_HAS_MONITOR_NAME;
+    info->srgb = !(output->flags & WAYLAND_OUTPUT_PRIMARIES);
+    info->preferred_width = output->current_mode->width;
+    info->preferred_height = output->current_mode->height;
+    info->preferred_refresh = output->current_mode->refresh / 1000.0;
+    info->width_mm = output->physical_w;
+    info->height_mm = output->physical_h;
+    ascii_to_unicode(info->monitor_name, output->model, 13);
+
+    if (output->flags & WAYLAND_OUTPUT_PRIMARIES)
+    {
+        info->r_x = output->primaries.r_x;
+        info->r_y = output->primaries.r_y;
+        info->g_x = output->primaries.g_x;
+        info->g_y = output->primaries.g_y;
+        info->b_x = output->primaries.b_x;
+        info->b_y = output->primaries.b_y;
+        info->w_x = output->primaries.w_x;
+        info->w_y = output->primaries.w_y;
+    }
+
+    if ((output->flags & cta_mask) == cta_mask && monitor.hdr_enabled)
+    {
+        info->max_fall = output->max_fall;
+        info->max_cll = output->max_cll;
+        info->flags |= MONITOR_INFO_HAS_CTA861_EXT;
+    }
+
     if ((env = getenv("DXVK_HDR")) && *env == '1')
         monitor.hdr_enabled = TRUE;
     else if ((env = getenv("DXVK_NO_HDR")) && *env == '1')
diff --git a/dlls/winewayland.drv/wayland_output.c b/dlls/winewayland.drv/wayland_output.c
index 26846b36644..568875ca519 100644
--- a/dlls/winewayland.drv/wayland_output.c
+++ b/dlls/winewayland.drv/wayland_output.c
@@ -35,16 +35,6 @@ WINE_DEFAULT_DEBUG_CHANNEL(waylanddrv);
 static const int32_t default_refresh = 60000;
 static uint32_t next_output_id = 0;
 
-#define WAYLAND_OUTPUT_CHANGED_MODES        0x01
-#define WAYLAND_OUTPUT_CHANGED_NAME         0x02
-#define WAYLAND_OUTPUT_CHANGED_LOGICAL_XY   0x04
-#define WAYLAND_OUTPUT_CHANGED_LOGICAL_WH   0x08
-#define WAYLAND_OUTPUT_CHANGED_GEOMETRY     0x10
-#define WAYLAND_OUTPUT_CHANGED_PRIMARIES    0x20
-#define WAYLAND_OUTPUT_CHANGED_FALL         0x40
-#define WAYLAND_OUTPUT_CHANGED_CLL          0x80
-#define WAYLAND_OUTPUT_CHANGED_LUMINANCES   0x100
-
 /**********************************************************************
  *          Output handling
  */
@@ -140,7 +130,7 @@ static void wayland_output_done(struct wayland_output *output)
     /* Update current state from pending state. */
     pthread_mutex_lock(&process_wayland.output_mutex);
 
-    if (output->pending_flags & WAYLAND_OUTPUT_CHANGED_GEOMETRY)
+    if (output->pending.flags & WAYLAND_OUTPUT_GEOMETRY)
     {
         free(output->current.make);
         free(output->current.model);
@@ -151,7 +141,7 @@ static void wayland_output_done(struct wayland_output *output)
         output->current.model = output->pending.model;
     }
 
-    if (output->pending_flags & WAYLAND_OUTPUT_CHANGED_MODES)
+    if (output->pending.flags & WAYLAND_OUTPUT_MODES)
     {
         RB_FOR_EACH_ENTRY(mode, &output->pending.modes, struct wayland_output_mode, entry)
         {
@@ -171,41 +161,35 @@ static void wayland_output_done(struct wayland_output *output)
         output->pending.modes_count = 0;
     }
 
-    if (output->pending_flags & WAYLAND_OUTPUT_CHANGED_NAME)
+    if (output->pending.flags & WAYLAND_OUTPUT_NAME)
     {
         free(output->current.name);
         output->current.name = output->pending.name;
         output->pending.name = NULL;
     }
 
-    if (output->pending_flags & WAYLAND_OUTPUT_CHANGED_LOGICAL_XY)
+    if (output->pending.flags & WAYLAND_OUTPUT_LOGICAL_XY)
     {
         output->current.logical_x = output->pending.logical_x;
         output->current.logical_y = output->pending.logical_y;
     }
 
-    if (output->pending_flags & WAYLAND_OUTPUT_CHANGED_LOGICAL_WH)
+    if (output->pending.flags & WAYLAND_OUTPUT_LOGICAL_WH)
     {
         output->current.logical_w = output->pending.logical_w;
         output->current.logical_h = output->pending.logical_h;
     }
 
-    if (output->pending_flags & WAYLAND_OUTPUT_CHANGED_PRIMARIES)
-    {
+    if (output->pending.flags & WAYLAND_OUTPUT_PRIMARIES)
         output->current.primaries = output->pending.primaries;
-    }
 
-    if (output->pending_flags & WAYLAND_OUTPUT_CHANGED_FALL)
-    {
+    if (output->pending.flags & WAYLAND_OUTPUT_FALL)
         output->current.max_fall = output->pending.max_fall;
-    }
 
-    if (output->pending_flags & WAYLAND_OUTPUT_CHANGED_CLL)
-    {
+    if (output->pending.flags & WAYLAND_OUTPUT_CLL)
         output->current.max_cll = output->pending.max_cll;
-    }
 
-    if (output->pending_flags & WAYLAND_OUTPUT_CHANGED_LUMINANCES)
+    if (output->pending.flags & WAYLAND_OUTPUT_LUMINANCES)
     {
         output->current.max_lum = output->pending.max_lum;
         output->current.ref_lum = output->pending.ref_lum;
@@ -214,7 +198,8 @@ static void wayland_output_done(struct wayland_output *output)
     output->current.supports_hdr = process_wayland.supports_win_scrgb &&
                                     (output->current.max_lum > output->current.ref_lum);
 
-    output->pending_flags = 0;
+    output->current.flags |= output->pending.flags;
+    output->pending.flags = 0;
 
     /* Ensure the logical dimensions have sane values. */
     if ((!output->current.logical_w || !output->current.logical_h) &&
@@ -222,6 +207,7 @@ static void wayland_output_done(struct wayland_output *output)
     {
         output->current.logical_w = output->current.current_mode->width;
         output->current.logical_h = output->current.current_mode->height;
+        output->current.flags |= WAYLAND_OUTPUT_LOGICAL_WH;
     }
 
     wayland_output_array_arrange_physical_coords();
@@ -257,7 +243,7 @@ static void output_handle_geometry(void *data, struct wl_output *wl_output,
     output->pending.model = strdup(model);
     output->pending.make = strdup(make);
 
-    output->pending_flags |= WAYLAND_OUTPUT_CHANGED_GEOMETRY;
+    output->pending.flags |= WAYLAND_OUTPUT_GEOMETRY;
 }
 
 static void output_handle_mode(void *data, struct wl_output *wl_output,
@@ -272,7 +258,7 @@ static void output_handle_mode(void *data, struct wl_output *wl_output,
     wayland_output_state_add_mode(&output->pending, width, height, refresh,
                                   (flags & WL_OUTPUT_MODE_CURRENT));
 
-    output->pending_flags |= WAYLAND_OUTPUT_CHANGED_MODES;
+    output->pending.flags |= WAYLAND_OUTPUT_MODES;
 }
 
 static void output_handle_done(void *data, struct wl_output *wl_output)
@@ -307,7 +293,7 @@ static void zxdg_output_v1_handle_logical_position(void *data,
     TRACE("logical_x=%d logical_y=%d\n", x, y);
     output->pending.logical_x = x;
     output->pending.logical_y = y;
-    output->pending_flags |= WAYLAND_OUTPUT_CHANGED_LOGICAL_XY;
+    output->pending.flags |= WAYLAND_OUTPUT_LOGICAL_XY;
 }
 
 static void zxdg_output_v1_handle_logical_size(void *data,
@@ -319,7 +305,7 @@ static void zxdg_output_v1_handle_logical_size(void *data,
     TRACE("logical_w=%d logical_h=%d\n", width, height);
     output->pending.logical_w = width;
     output->pending.logical_h = height;
-    output->pending_flags |= WAYLAND_OUTPUT_CHANGED_LOGICAL_WH;
+    output->pending.flags |= WAYLAND_OUTPUT_LOGICAL_WH;
 }
 
 static void zxdg_output_v1_handle_done(void *data,
@@ -340,7 +326,7 @@ static void zxdg_output_v1_handle_name(void *data,
 
     free(output->pending.name);
     output->pending.name = strdup(name);
-    output->pending_flags |= WAYLAND_OUTPUT_CHANGED_NAME;
+    output->pending.flags |= WAYLAND_OUTPUT_NAME;
 }
 
 static void zxdg_output_v1_handle_description(void *data,
@@ -398,7 +384,7 @@ static void wayland_image_description_info_v1_luminance(void *data,
 
     output->pending.ref_lum = ref;
     output->pending.max_lum = max;
-    output->pending_flags |= WAYLAND_OUTPUT_CHANGED_LUMINANCES;
+    output->pending.flags |= WAYLAND_OUTPUT_LUMINANCES;
 }
 
 static void wayland_image_description_info_v1_primaries(void *data,
@@ -431,7 +417,7 @@ static void wayland_image_description_info_v1_target_primaries(void *data,
             r_x * 1e-6, r_y * 1e-6, g_x * 1e-6, g_y * 1e-6,
             b_x * 1e-6, b_y * 1e-6, w_x * 1e-6, w_y * 1e-6);
 
-    output->pending_flags |= WAYLAND_OUTPUT_CHANGED_PRIMARIES;
+    output->pending.flags |= WAYLAND_OUTPUT_PRIMARIES;
 }
 
 static void wayland_image_description_info_v1_target_luminance(void *data,
@@ -449,7 +435,7 @@ static void wayland_image_description_info_v1_target_max_cll(void *data,
     TRACE("Max CLL: %u\n", max);
 
     output->pending.max_cll = max;
-    output->pending_flags |= WAYLAND_OUTPUT_CHANGED_CLL;
+    output->pending.flags |= WAYLAND_OUTPUT_CLL;
 }
 
 static void wayland_image_description_info_v1_target_max_fall(void *data,
@@ -460,7 +446,7 @@ static void wayland_image_description_info_v1_target_max_fall(void *data,
     TRACE("Max FALL: %u\n", max);
 
     output->pending.max_fall = max;
-    output->pending_flags |= WAYLAND_OUTPUT_CHANGED_FALL;
+    output->pending.flags |= WAYLAND_OUTPUT_FALL;
 }
 
 static const struct wp_image_description_info_v1_listener image_description_info_listener = {
diff --git a/dlls/winewayland.drv/waylanddrv.h b/dlls/winewayland.drv/waylanddrv.h
index 4b5d8abd3f3..6e4cf432da0 100644
--- a/dlls/winewayland.drv/waylanddrv.h
+++ b/dlls/winewayland.drv/waylanddrv.h
@@ -281,6 +281,16 @@ struct wayland_output_mode
     int32_t refresh;
 };
 
+#define WAYLAND_OUTPUT_MODES        0x01
+#define WAYLAND_OUTPUT_NAME         0x02
+#define WAYLAND_OUTPUT_LOGICAL_XY   0x04
+#define WAYLAND_OUTPUT_LOGICAL_WH   0x08
+#define WAYLAND_OUTPUT_GEOMETRY     0x10
+#define WAYLAND_OUTPUT_PRIMARIES    0x20
+#define WAYLAND_OUTPUT_FALL         0x40
+#define WAYLAND_OUTPUT_CLL          0x80
+#define WAYLAND_OUTPUT_LUMINANCES   0x100
+
 struct wayland_primaries
 {
     int32_t r_x;
@@ -312,6 +322,7 @@ struct wayland_output_state
     uint32_t max_lum;
     uint32_t ref_lum;
     BOOL supports_hdr;
+    unsigned int flags;
 };
 
 struct wayland_output
@@ -323,7 +334,6 @@ struct wayland_output
     struct wp_image_description_info_v1 *wp_image_description_info_v1;
     struct wp_color_management_output_v1 *wp_color_management_output_v1;
     uint32_t global_id;
-    unsigned int pending_flags;
     LONG ref;
     struct wayland_output_state pending, current;
 };
-- 
2.55.0

